File: cortexdb-js/src/ingest.ts
/**
 * Ingest API wrapper for CortexDB.
 *
 * Equivalent to cortexdb-py/cortexdb/ingest.py
 */

import { HttpClient } from "./httpWrapper.js";
import {
    ConverserRole,
    type IngestPromptRequest,
    type IngestDocumentRequest,
    type IngestResponse,
} from "./models.js";

export class IngestAPI {
    /** @internal */
    constructor(private readonly http: HttpClient) { }

    /**
     * Ingest a prompt payload into CortexDB.
     * The server will perform semantic compression and online synthesis.
     *
     * @param uid        User identifier.
     * @param converser  Role of the converser ("USER", "AGENT", or "SYSTEM").
     * @param text       The text content to ingest.
     * @param metadata   Optional metadata dictionary.
     * @returns IngestResponse with the created KnowledgeBase and processing info.
     */
    async prompt(
        uid: string,
        converser: string | ConverserRole,
        text: string,
        metadata?: Record<string, any>,
    ): Promise<IngestResponse> {
        // Normalise string converser to enum value
        const resolvedConverser =
            typeof converser === "string"
                ? (converser.toUpperCase() as ConverserRole)
                : converser;

        const request: IngestPromptRequest = {
            uid,
            converser: resolvedConverser,
            text,
            ...(metadata !== undefined && { metadata }),
        };

        return this.http.post<IngestResponse>("/api/v1/memory/ingest/prompt", request);
    }

    /**
     * Ingest a large document into CortexDB.
     * The server will extract a hierarchical page index (document tree).
     *
     * @param uid           User identifier.
     * @param documentTitle Title of the document.
     * @param documentText  The full text content of the document.
     * @returns IngestResponse with the created KnowledgeBase and processing info.
     */
    async document(
        uid: string,
        documentTitle: string,
        documentText: string,
    ): Promise<IngestResponse> {
        const request: IngestDocumentRequest = {
            uid,
            documentTitle,
            documentText,
        };

        return this.http.post<IngestResponse>("/api/v1/memory/ingest/document", request);
    }
}

File: cortexdb-py/cortexdb/ingest.py
"""Ingest API wrapper for CortexDB."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from cortexdb.models import ConverserRole, IngestPromptRequest, IngestDocumentRequest, IngestResponse

if TYPE_CHECKING:
    import httpx


class IngestAPI:
    """Wraps the ``/api/v1/memory/ingest`` endpoints."""

    def __init__(self, http: httpx.Client) -> None:
        self._http = http

    def prompt(
        self,
        uid: str,
        converser: str | ConverserRole,
        text: str,
        metadata: dict[str, Any] | None = None,
    ) -> IngestResponse:
        """Ingest a prompt payload into CortexDB.

        The server will perform semantic compression and online synthesis.

        Args:
            uid: User identifier.
            converser: Role of the converser (USER, AGENT, or SYSTEM).
            text: The text content to ingest.
            metadata: Optional metadata dict.

        Returns:
            IngestResponse with processing info.
        """
        if isinstance(converser, str):
            converser = ConverserRole(converser.upper())

        request = IngestPromptRequest(
            uid=uid,
            converser=converser,
            text=text,
            metadata=metadata,
        )
        response = self._http.post(
            "/api/v1/memory/ingest/prompt",
            json=request.model_dump(exclude_none=True),
        )
        response.raise_for_status()
        return IngestResponse.model_validate(response.json())

    def document(
        self,
        uid: str,
        document_title: str,
        document_text: str,
    ) -> IngestResponse:
        """Ingest a large document into CortexDB.

        The server will extract a hierarchical page index (document tree).

        Args:
            uid: User identifier.
            document_title: Title of the document.
            document_text: The full text content to ingest.

        Returns:
            IngestResponse with processing info.
        """
        request = IngestDocumentRequest(
            uid=uid,
            document_title=document_title,
            document_text=document_text,
        )
        response = self._http.post(
            "/api/v1/memory/ingest/document",
            json=request.model_dump(by_alias=True, exclude_none=True),
        )
        response.raise_for_status()
        return IngestResponse.model_validate(response.json())
